test/pkg/environment/common/expectations.go

```diff
@@ -565,34 +565,73 @@ func (env *Environment) ConsistentlyExpectNodeCount(comparator string, count int
+		g.Expect(nodeClaimList.Items).To(HaveLen(nodeCount))
+		nodeList := &corev1.NodeList{}
+		g.Expect(env.Client.List(env, nodeList, client.HasLabels{test.DiscoveryLabel})).To(Succeed())
+		g.Expect(nodeList.Items).To(HaveLen(nodeCount))
+		nodeList.Items = lo.Filter(nodeList.Items, func(n corev1.Node, _ int) bool {
+			_, ok := lo.Find(n.Spec.Taints, func(t corev1.Taint) bool {
+				return t.MatchTaint(&karpv1.DisruptedNoScheduleTaint)
+			})
+			return ok
+		})
+		g.Expect(nodeList.Items).To(HaveLen(0))
+	}, duration).Should(Succeed())
 }
 
-// ConsistentlyExpectDisruptionsWithNodeCount will continually ensure that there are exactly disruptingNodes with totalNodes (including replacements and existing nodes)
-func (env *Environment) ConsistentlyExpectDisruptionsWithNodeCount(disruptingNodes, totalNodes int, duration time.Duration) (taintedNodes []*corev1.Node) {
+// ConsistentlyExpectDisruptionsWithNodeCount consistently ensures a max on number of concurrently disrupting and non-terminating nodes.
+// For example: if we have 5 nodes, with a budget of 2 nodes, we ensure that `disruptingNodes <= 2`
+// We use nonDisrupting to assert that we're not creating too many instances in replacement.
+func (env *Environment) ConsistentlyExpectDisruptionsUntilTarget(disruptingNodes, startingPoint, target int, timeout time.Duration) {
 	GinkgoHelper()
 	nodes := []corev1.Node{}
-	Consistently(func(g Gomega) {
+	// We use an eventually to exit when we detect the number of tainted/disrupted nodes matches our target.
+	Eventually(func(g Gomega) {
 		// Ensure we don't change our NodeClaims
 		nodeClaimList := &karpv1.NodeClaimList{}
 		g.Expect(env.Client.List(env, nodeClaimList, client.HasLabels{test.DiscoveryLabel})).To(Succeed())
-		g.Expect(nodeClaimList.Items).To(HaveLen(totalNodes))
+		// We don't consider nodes that have the `terminating` status condition towards our budget, so we shouldn't consider these nodes in our budget.
+		removedProviderIDs := sets.Set[string]{}
+		nodeClaimList.Items = lo.Filter(nodeClaimList.Items, func(nc karpv1.NodeClaim, _ int) bool {
+			if !nc.StatusConditions().IsTrue(karpv1.ConditionTypeInstanceTerminating) {
+				return true
+			}
+			removedProviderIDs.Insert(nc.Status.ProviderID)
+			return false
+		})
+		if len(nodeClaimList.Items) > startingPoint+disruptingNodes {
+			StopTrying(fmt.Sprintf("Too many nodeclaims created. Expected no more than %d, got %d", startingPoint+disruptingNodes, len(nodeClaimList.Items))).Now()
+		}
 
 		nodeList := &corev1.NodeList{}
 		g.Expect(env.Client.List(env, nodeList, client.HasLabels{test.DiscoveryLabel})).To(Succeed())
-		g.Expect(nodeList.Items).To(HaveLen(totalNodes))
+		nodeList.Items = lo.Filter(nodeList.Items, func(n corev1.Node, _ int) bool {
+			return !removedProviderIDs.Has(n.Spec.ProviderID)
+		})
+		g.Expect(len(nodeList.Items)).To(BeNumerically("<=", startingPoint+disruptingNodes))
```

Comment:
Aren't these two checks reduant? If one bails, I don't think you need the other one?

Reply:
Good call